{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'o': 'b', 10: 2, 'b': False}\n"
     ]
    }
   ],
   "source": [
    "d1 = {'a': 'b', 5: 2, True: False} \n",
    "\n",
    "d2 = {'o': 'a', 10: 5, 'b': True} \n",
    "\n",
    "d3 = {} \n",
    "\n",
    "for k2, v2 in d2.items(): \n",
    "\n",
    "    for k1, v1 in d1.items(): \n",
    "\n",
    "        if v2 == k1: \n",
    "\n",
    "            d3[k2] = v1 \n",
    "\n",
    "print (d3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'o': {'a': 'b'}, 10: {5: 2}, 'b': {True: False}}\n"
     ]
    }
   ],
   "source": [
    "d1 = {'a': 'b', 5: 2, True: False} \n",
    "\n",
    "d2 = {'o': 'a', 10: 5, 'b': True} \n",
    "\n",
    "d3 = {} \n",
    "\n",
    "for k2, v2 in d2.items():\n",
    "    d3[k2]={}\n",
    "    for k1, v1 in d1.items(): \n",
    "        if v2 == k1: \n",
    "            d3[k2][v2] = v1 \n",
    "print(d3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'A': {'a': '11', 'b': '22', 'c': '33'}, 'B': {'d': '44', 'e': '55', 'f': '66'}}\n",
      "0.0003056154546356993\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "starttime=time.clock()\n",
    "a1={'A':['a','b','c'],'B':['d','e','f']}\n",
    "a2={'a':'11','b':'22','c':'33','d':'44','e':'55','f':'66'}\n",
    "a3={}\n",
    "for k1,v1 in a1.items():\n",
    "    a3[k1]={}\n",
    "    for i in v1:\n",
    "        for k2,v2 in a2.items():\n",
    "            if i==k2:\n",
    "                a3[k1][i]=v2\n",
    "print(a3)\n",
    "endtime=time.clock()\n",
    "print(endtime - starttime)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'A': ['a\\t11', 'b\\t22', 'c\\t33'], 'B': ['d\\t44', 'e\\t55', 'f\\t66']}\n",
      "0.0003588320776088949\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "starttime=time.clock()\n",
    "a1={'A':['a','b','c'],'B':['d','e','f']}\n",
    "a2={'a':'11','b':'22','c':'33','d':'44','e':'55','f':'66'}\n",
    "a3={}\n",
    "for k1,v1 in a1.items():\n",
    "    a3[k1]=[]\n",
    "    for i in v1:\n",
    "        for k2,v2 in a2.items():\n",
    "            if i==k2:\n",
    "                a3[k1].append(i+\"\\t\"+v2)\n",
    "print(a3)\n",
    "endtime=time.clock()\n",
    "print(endtime - starttime)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'A': {'a': '11', 'b': '22', 'c': '33'}, 'B': {'d': '44', 'e': '55', 'f': '66'}}\n",
      "0.00032196056099564885\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "starttime=time.clock()\n",
    "a1={'A':['a','b','c'],'B':['d','e','f']}\n",
    "a2={'a':'11','b':'22','c':'33','d':'44','e':'55','f':'66'}\n",
    "a3={}\n",
    "for k1,v1 in a1.items():\n",
    "    a3[k1]={}\n",
    "    for i in v1:\n",
    "        if i in a2:\n",
    "            a3[k1][i]=a2[i]\n",
    "print(a3)\n",
    "endtime=time.clock()\n",
    "print(endtime - starttime)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'A': {'a': '11', 'b': '22', 'c': '33'}, 'B': {'d': '44', 'e': '55', 'f': '66'}}\n",
      "0.0002717848892643815\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "starttime=time.clock()\n",
    "a1={'A':['a','b','c'],'B':['d','e','f']}\n",
    "a2={'a':'11','b':'22','c':'33','d':'44','e':'55','f':'66'}\n",
    "a3={}\n",
    "for k1,v1 in a1.items():\n",
    "    a3[k1]={}\n",
    "    for i in v1:\n",
    "        a3[k1][i]=a2[i]\n",
    "print(a3)\n",
    "endtime=time.clock()\n",
    "print(endtime - starttime)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "基因——转录本——外显子坐标统计\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import re\n",
    " \n",
    "args = sys.argv    #实现从程序外部向程序传递参数\n",
    " \n",
    " \n",
    "class Genome_info:\n",
    "    def __init__(self):\n",
    "        self.chr = \"\"\n",
    "        self.start = 0\n",
    "        self.end = 0\n",
    " \n",
    " \n",
    "class Gene(Genome_info):\n",
    "    def __init__(self):\n",
    "        Genome_info.__init__(self)\n",
    "        self.orientation = \"\"\n",
    "        self.id = \"\"\n",
    "        self.name = \"\"    #这个是我加的\n",
    "        self.catag = \"\"   #这个也是我加的\n",
    " \n",
    " \n",
    "class Transcript(Genome_info):\n",
    "    def __init__(self):\n",
    "        Genome_info.__init__(self)\n",
    "        self.id = \"\"\n",
    "        self.parent = \"\"\n",
    " \n",
    " \n",
    "class Exon(Genome_info):\n",
    "    def __init__(self):\n",
    "        Genome_info.__init__(self)\n",
    "        self.parent = \"\"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def main(args):\n",
    "    \"\"\"\n",
    "    一个输入参数：\n",
    "    第一个参数为物种gtf文件\n",
    " \n",
    "    :return:\n",
    "    \"\"\"\n",
    " \n",
    "    list_chr = []\n",
    "    list_gene = {}\n",
    "    list_transcript = {}\n",
    "    list_exon = []\n",
    "    # l_n = 0\n",
    "    with open(args[1]) as fp_gtf:\n",
    "        for line in fp_gtf:\n",
    "            if line.startswith(\"#\"):\n",
    "                continue\n",
    "            # print (\"in %s\" % l_n)\n",
    "            # l_n += 1\n",
    "            lines = line.strip(\"\\n\").split(\"\\t\")   #去掉换行符，以tab健进行切片\n",
    "            chr = lines[0]\n",
    "            type = lines[2]                        #gene,transcript,exon,CDS,start_codon,stop_codon,five_prime_utr,three_prime_utr,Selenocysteine\n",
    "            start = int(lines[3])\n",
    "            end = int(lines[4])\n",
    "            orientation = lines[6]\n",
    "            attr = lines[8]\n",
    "            #if not re.search(r'protein_coding', attr):     #不要没有protein_coding的行\n",
    "                #continue\n",
    " \n",
    "            if not chr in list_chr :                       #把染色体号加入列表中\n",
    "                list_chr.append(chr)\n",
    " \n",
    "            if type == \"gene\":                             #如果类型为基因\n",
    "                gene = Gene()                              #初始化一个基因的对象，Gene是一个类，gene是Gene类的一个实例，这个实例有5个属性\n",
    "                id = re.search(r'gene_id \"([^;]+)\";?', attr).group(1)   #在属性attr里边匹配gene_id,以gene_id,以0个或者1个;结尾，中间匹配至少一个非;字符\n",
    "                name = re.search(r'gene_name \"([^;]+)\";?', attr).group(1)\n",
    "                catag = re.search(r'gene_biotype \"([^;]+)\";?', attr).group(1)   #对基因的类型进行分类\n",
    "                gene.chr = chr\n",
    "                gene.start = start\n",
    "                gene.end = end\n",
    "                gene.id = id\n",
    "                gene.name = name\n",
    "                gene.catag = catag\n",
    "                gene.orientation = orientation   #chr,start,end,orientation是通过切片直接获得，而id通过正则表达式获得  \n",
    "                list_gene[id] = gene             #以gene_id为健，有5个属性的gene为值放入名为list_gene的字典\n",
    "                # print(id)\n",
    "            elif type == \"transcript\":\n",
    "                transcript = Transcript()\n",
    "                id = re.search(r'transcript_id \"([^;]+)\";?', attr).group(1)\n",
    "                parent = re.search(r'gene_id \"([^;]+)\";?', attr).group(1)          #这里的gene_id就成了transcript的parent\n",
    "                if not parent in list_gene:\n",
    "                    continue\n",
    "                transcript.chr = chr\n",
    "                transcript.start = start\n",
    "                transcript.end = end\n",
    "                transcript.id = id\n",
    "                transcript.parent = parent\n",
    "                list_transcript[id] = transcript                              #\n",
    " \n",
    "            elif type == \"exon\":\n",
    "                exon = Exon()\n",
    "                parent = re.search(r'transcript_id \"([^;]+)\";?', attr).group(1)\n",
    "                if not parent in list_transcript:\n",
    "                    continue\n",
    "                exon.chr = chr\n",
    "                exon.start = start\n",
    "                exon.end = end\n",
    "                exon.parent = parent\n",
    "                list_exon.append(exon)                                         #\n",
    " \n",
    "    #chr_gene(list_gene)\n",
    "    #gene_len(list_gene)\n",
    "    #gene_transcript(list_transcript)\n",
    "    #transcript_exon(list_exon)\n",
    "    #exon_pos(list_exon)\n",
    "    gene_exon_pos(list_gene, list_transcript, list_exon)\n",
    "def gene_exon_pos(list_gene, list_transcript, list_exon):\n",
    "    \"\"\"\n",
    "    根据exon的parent将所有exon对应到transcript\n",
    "    根据transcript的parent将所有transcript对应到gene\n",
    "    根据gene按chr分组得到chromosome列表\n",
    " \n",
    "    从chromosome中输出某个指定基因的所有外显子坐标信息并画图\n",
    "    生信编程直播第五题\n",
    " \n",
    "    :param list_gene:\n",
    "    :param list_transcript:\n",
    "    :param list_exon:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    print(\"基因——转录本——外显子坐标统计\")\n",
    "    gene_transcript_exon = {}\n",
    "    count_exon = {}\n",
    "    count_transcript= {}\n",
    "    for exon in list_exon:\n",
    "        transcript_id = exon.parent\n",
    "        if transcript_id in count_exon:\n",
    "            count_exon[transcript_id] += \",%s-%s\" % (str(exon.start), str(exon.end))\n",
    "        else:\n",
    "            count_exon[transcript_id] = \"%s-%s\" % (str(exon.start), str(exon.end))\n",
    "    for transcript_id,transcript in list_transcript.items():\n",
    "        gene_id = transcript.parent\n",
    "        if gene_id in count_transcript:\n",
    "            count_transcript[gene_id].append(transcript_id)\n",
    "        else:\n",
    "            count_transcript[gene_id]=[]\n",
    "    #print(count_transcript)\n",
    "    for k1,v1 in count_transcript.items():\n",
    "        gene_transcript_exon[k1]={}\n",
    "        for i in v1:\n",
    "            gene_transcript_exon[k1][i]=count_exon[i]\n",
    "    #print(gene_transcript_exon)        \n",
    "    with open(\"gene_transcript_exon.txt\", 'w') as fp_out:\n",
    "        for gene_id, transcript in gene_transcript_exon.items():\n",
    "            for transcript_id,exon_pos in transcript.items():\n",
    "            #print(\"\\t\".join([transcript_id, pos]) + \"\\n\")\n",
    "                fp_out.write(\"\\t\".join([gene_id,transcript_id,exon_pos]) + \"\\n\")\n",
    "\n",
    "args=['1','Homo_sapiens.GRCh38.87.chr.gtf']\n",
    "#args=['1','test2.txt']\n",
    "if __name__ == \"__main__\":\n",
    "    main(args)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "基因——转录本——外显子坐标统计\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import re\n",
    " \n",
    "args = sys.argv    #实现从程序外部向程序传递参数\n",
    " \n",
    " \n",
    "class Genome_info:\n",
    "    def __init__(self):\n",
    "        self.chr = \"\"\n",
    "        self.start = 0\n",
    "        self.end = 0\n",
    " \n",
    " \n",
    "class Gene(Genome_info):\n",
    "    def __init__(self):\n",
    "        Genome_info.__init__(self)\n",
    "        self.orientation = \"\"\n",
    "        self.id = \"\"\n",
    "        self.name = \"\"    #这个是我加的\n",
    "        self.catag = \"\"   #这个也是我加的\n",
    " \n",
    " \n",
    "class Transcript(Genome_info):\n",
    "    def __init__(self):\n",
    "        Genome_info.__init__(self)\n",
    "        self.id = \"\"\n",
    "        self.parent = \"\"\n",
    " \n",
    " \n",
    "class Exon(Genome_info):\n",
    "    def __init__(self):\n",
    "        Genome_info.__init__(self)\n",
    "        self.parent = \"\"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def main(args):\n",
    "    \"\"\"\n",
    "    一个输入参数：\n",
    "    第一个参数为物种gtf文件\n",
    " \n",
    "    :return:\n",
    "    \"\"\"\n",
    " \n",
    "    list_chr = []\n",
    "    list_gene = {}\n",
    "    list_transcript = {}\n",
    "    list_exon = []\n",
    "    # l_n = 0\n",
    "    with open(args[1]) as fp_gtf:\n",
    "        for line in fp_gtf:\n",
    "            if line.startswith(\"#\"):\n",
    "                continue\n",
    "            # print (\"in %s\" % l_n)\n",
    "            # l_n += 1\n",
    "            lines = line.strip(\"\\n\").split(\"\\t\")   #去掉换行符，以tab健进行切片\n",
    "            chr = lines[0]\n",
    "            type = lines[2]                        #gene,transcript,exon,CDS,start_codon,stop_codon,five_prime_utr,three_prime_utr,Selenocysteine\n",
    "            start = int(lines[3])\n",
    "            end = int(lines[4])\n",
    "            orientation = lines[6]\n",
    "            attr = lines[8]\n",
    "            #if not re.search(r'protein_coding', attr):     #不要没有protein_coding的行\n",
    "                #continue\n",
    " \n",
    "            if not chr in list_chr :                       #把染色体号加入列表中\n",
    "                list_chr.append(chr)\n",
    " \n",
    "            if type == \"gene\":                             #如果类型为基因\n",
    "                gene = Gene()                              #初始化一个基因的对象，Gene是一个类，gene是Gene类的一个实例，这个实例有5个属性\n",
    "                id = re.search(r'gene_id \"([^;]+)\";?', attr).group(1)   #在属性attr里边匹配gene_id,以gene_id,以0个或者1个;结尾，中间匹配至少一个非;字符\n",
    "                name = re.search(r'gene_name \"([^;]+)\";?', attr).group(1)\n",
    "                catag = re.search(r'gene_biotype \"([^;]+)\";?', attr).group(1)   #对基因的类型进行分类\n",
    "                gene.chr = chr\n",
    "                gene.start = start\n",
    "                gene.end = end\n",
    "                gene.id = id\n",
    "                gene.name = name\n",
    "                gene.catag = catag\n",
    "                gene.orientation = orientation   #chr,start,end,orientation是通过切片直接获得，而id通过正则表达式获得  \n",
    "                list_gene[id] = gene             #以gene_id为健，有5个属性的gene为值放入名为list_gene的字典\n",
    "                # print(id)\n",
    "            elif type == \"transcript\":\n",
    "                transcript = Transcript()\n",
    "                id = re.search(r'transcript_id \"([^;]+)\";?', attr).group(1)\n",
    "                parent = re.search(r'gene_id \"([^;]+)\";?', attr).group(1)          #这里的gene_id就成了transcript的parent\n",
    "                if not parent in list_gene:\n",
    "                    continue\n",
    "                transcript.chr = chr\n",
    "                transcript.start = start\n",
    "                transcript.end = end\n",
    "                transcript.id = id\n",
    "                transcript.parent = parent\n",
    "                list_transcript[id] = transcript                              #\n",
    " \n",
    "            elif type == \"exon\":\n",
    "                exon = Exon()\n",
    "                parent = re.search(r'transcript_id \"([^;]+)\";?', attr).group(1)\n",
    "                if not parent in list_transcript:\n",
    "                    continue\n",
    "                exon.chr = chr\n",
    "                exon.start = start\n",
    "                exon.end = end\n",
    "                exon.parent = parent\n",
    "                list_exon.append(exon)                                         #\n",
    " \n",
    "    #chr_gene(list_gene)\n",
    "    #gene_len(list_gene)\n",
    "    #gene_transcript(list_transcript)\n",
    "    #transcript_exon(list_exon)\n",
    "    #exon_pos(list_exon)\n",
    "    gene_exon_pos(list_gene, list_transcript, list_exon)\n",
    "def gene_exon_pos(list_gene, list_transcript, list_exon):\n",
    "    \"\"\"\n",
    "    根据exon的parent将所有exon对应到transcript\n",
    "    根据transcript的parent将所有transcript对应到gene\n",
    "    根据gene按chr分组得到chromosome列表\n",
    " \n",
    "    从chromosome中输出某个指定基因的所有外显子坐标信息并画图\n",
    "    生信编程直播第五题\n",
    " \n",
    "    :param list_gene:\n",
    "    :param list_transcript:\n",
    "    :param list_exon:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    print(\"基因——转录本——外显子坐标统计\")\n",
    "    gene_transcript_exon = {}\n",
    "    count_exon = {}\n",
    "    count_transcript= {}\n",
    "    for exon in list_exon:\n",
    "        transcript_id = exon.parent\n",
    "        if transcript_id in count_exon:\n",
    "            count_exon[transcript_id] += \",%s-%s\" % (str(exon.start), str(exon.end))\n",
    "        else:\n",
    "            count_exon[transcript_id] = \"%s-%s\" % (str(exon.start), str(exon.end))\n",
    "    for transcript_id,transcript in list_transcript.items():\n",
    "        gene_id = transcript.parent\n",
    "        if gene_id in count_transcript:\n",
    "            count_transcript[gene_id].append(transcript_id)\n",
    "        else:\n",
    "            count_transcript[gene_id]=[]\n",
    "    #print(count_transcript)\n",
    "    for k1,v1 in count_transcript.items():\n",
    "        gene_transcript_exon[k1]={}\n",
    "        for i in v1:\n",
    "            if i in count_exon:\n",
    "                gene_transcript_exon[k1][i]=count_exon[i]\n",
    "    #print(gene_transcript_exon)        \n",
    "    with open(\"gene_transcript_exon1.txt\", 'w') as fp_out:\n",
    "        for gene_id, transcript in gene_transcript_exon.items():\n",
    "            for transcript_id,exon_pos in transcript.items():\n",
    "            #print(\"\\t\".join([transcript_id, pos]) + \"\\n\")\n",
    "                fp_out.write(\"\\t\".join([gene_id,transcript_id,exon_pos]) + \"\\n\")\n",
    "\n",
    "args=['1','Homo_sapiens.GRCh38.87.chr.gtf']\n",
    "#args=['1','test2.txt']\n",
    "if __name__ == \"__main__\":\n",
    "    main(args)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
